internal/testmain/testmain.go

```diff
@@ -0,0 +1,307 @@
+package testmain
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/ast"
+	gobuild "go/build"
+	"go/doc"
+	"go/parser"
+	"go/token"
+	"path"
+	"sort"
+	"strings"
+	"text/template"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/gopherjs/gopherjs/build"
+	"golang.org/x/tools/go/buildutil"
+)
+
+// FuncLocation describes whether a test function is in-package or external
+// (i.e. in the xxx_test package).
+type FuncLocation uint8
+
+const (
+	// LocUnknown is the default, invalid value of the PkgType.
+	LocUnknown FuncLocation = iota
+	// LocInPackage is an in-package test.
+	LocInPackage
+	// LocExternal is an external test (i.e. in the xxx_test package).
+	LocExternal
+)
+
+func (tl FuncLocation) String() string {
+	switch tl {
+	case LocInPackage:
+		return "_test"
+	case LocExternal:
+		return "_xtest"
+	default:
+		return "<unknown>"
+	}
+}
+
+// TestFunc describes a single test/benchmark/fuzz function in a package.
+type TestFunc struct {
+	Location FuncLocation // Where the function is defined.
+	Name     string       // Function name.
+}
+
+// ExampleFunc describes an example.
+type ExampleFunc struct {
+	Location    FuncLocation // Where the function is defined.
+	Name        string       // Function name.
+	Output      string       // Expected output.
+	Unordered   bool         // Output is allowed to be unordered.
+	EmptyOutput bool         // Whether the output is expected to be empty.
+}
+
+// Executable returns true if the example function should be executed with tests.
+func (ef ExampleFunc) Executable() bool {
+	return ef.EmptyOutput || ef.Output != ""
+}
+
+// TestMain is a helper type responsible for generation of the test main package.
+type TestMain struct {
+	Package    *build.PackageData
+	Tests      []TestFunc
+	Benchmarks []TestFunc
+	Fuzz       []TestFunc
+	Examples   []ExampleFunc
+	TestMain   *TestFunc
+}
+
+// Scan package for tests functions.
+func (tm *TestMain) Scan(fset *token.FileSet) error {
+	if err := tm.scanPkg(fset, tm.Package.TestGoFiles, LocInPackage); err != nil {
+		return err
+	}
+	if err := tm.scanPkg(fset, tm.Package.XTestGoFiles, LocExternal); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (tm *TestMain) scanPkg(fset *token.FileSet, files []string, loc FuncLocation) error {
+	for _, name := range files {
+		srcPath := path.Join(tm.Package.Dir, name)
+		f, err := buildutil.OpenFile(tm.Package.InternalBuildContext(), srcPath)
+		if err != nil {
+			return fmt.Errorf("failed to open source file %q: %w", srcPath, err)
+		}
+		defer f.Close()
```

Comment:
Nit: Since deferred functions don't execute until the entire function returns, we could end up with a large number of open files here.  Probably not likely to be an actual problem, unless some package has thousands of files in it. Do we care? :shrug:

Reply:
I don't think we do until proven otherwise. That would have to be a single package with *hundreds* of test source files (not just any source files), which I don't think I've seen. But if we get a bug report like that, I promise to fix this :)